### ISSUE TITLE
Multi GPU support is ... ornamental?

### ISSUE THREAD

**Issue by mbabadi** [Spijkervet/SimCLR#37]
Has the code been tested on multi GPUs in the DDP mode? -- not in the sense that it "runs" (it certainly does), but in the sense that it actually learns a representation (it likely doesn't)?

**Comment by mbabadi** [comment]
I believe there's a bug in `NT_Xent` .  As it stands, it doesn't learn a meaningful representation when `world_size > 1` despite the loss going down. There must be a subtle issue in concatenating `z_i` and `z_j` before gathering and then masking.

Since the code works for single GPU, I tried to mimic the single-GPU logic as follows. The following updated implementation of `NT_Xent` runs and produces good representations for both single- and multi- GPU:

```
import torch
import torch.distributed as dist
from .gather import GatherLayer


class NT_Xent(torch.nn.Module):
    
    def __init__(self, batch_size, temperature, world_size):
        super(NT_Xent, self).__init__()
        
        self.batch_size = batch_size
        self.temperature = temperature
        self.world_size = world_size
        self.full_batch_size = world_size * batch_size
        
        self.mask = self.get_self_excluding_mask()
        self.criterion = torch.nn.CrossEntropyLoss(reduction="sum")
        self.similarity_f = torch.nn.CosineSimilarity(dim=2)

    def get_self_excluding_mask(self) -> torch.Tensor:
        mask = torch.ones((2 * self.full_batch_size, 2 * self.full_batch_size), dtype=bool)
        mask = mask.fill_diagonal_(0)
        for i in range(self.full_batch_size):
            mask[i, i + self.full_batch_size] = 0
            mask[i + self.full_batch_size, i] = 0
        return mask

    def forward(self, z_i: torch.Tensor, z_j: torch.Tensor) -> torch.Tensor:

        if self.world_size > 1:
            z_i = torch.cat(GatherLayer.apply(z_i), dim=0)
            z_j = torch.cat(GatherLayer.apply(z_j), dim=0)
        
        # first M entries of z will be the first augmentation; last M entries will
        # be the second augmentation
        z = torch.cat((z_i, z_j), dim=0)
        
        # the dimension of sim is (2 * self.full_batch_size) x (2 * self.full_batch_size)
        sim = self.similarity_f(z.unsqueeze(1), z.unsqueeze(0)) / self.temperature

        sim_i_j = torch.diag(sim, self.full_batch_size)
        sim_j_i = torch.diag(sim, -self.full_batch_size)

        positive_samples = torch.cat((sim_i_j, sim_j_i), dim=0).reshape(2 * self.full_batch_size, 1)
        negative_samples = sim[self.mask].reshape(2 * self.full_batch_size, -1)

        labels = torch.zeros(2 * self.full_batch_size).to(positive_samples.device).long()
        logits = torch.cat((positive_samples, negative_samples), dim=1)
        loss = self.criterion(logits, labels) / (2 * self.full_batch_size)
        
        return loss
```

I have changed the order of `GatherLayer` and `torch.cat` operations on `z_i` and `z_j`;  I have also changed the positive-pair/self-pair mask.

**Comment by Spijkervet** [comment]
Hi there! This bug was addressed in https://github.com/Spijkervet/SimCLR/commit/cd85c4366d2e6ac1b0a16798b76ac0a2c8a94e58 - the masking for world_size > 1 wasn't done properly. This only affected results when using multi-GPU training with DistributedDataParallel.

**Comment by kevinbro96** [comment]
Yes

**Comment by chenxi52** [comment]
emmm, you didn't sove this problem unless you use:
z_i = torch.cat(GatherLayer.apply(z_i), dim=0)
 z_j = torch.cat(GatherLayer.apply(z_j), dim=0

> Hi there! This bug was addressed in [cd85c43](https://github.com/Spijkervet/SimCLR/commit/cd85c4366d2e6ac1b0a16798b76ac0a2c8a94e58) - the masking for world_size > 1 wasn't done properly. This only affected results when using multi-GPU training with DistributedDataParallel.

**Comment by JUNJIE99** [comment]
> I believe there's a bug in `NT_Xent` . As it stands, it doesn't learn a meaningful representation when `world_size > 1` despite the loss going down. There must be a subtle issue in concatenating `z_i` and `z_j` before gathering and then masking.
> 
> Since the code works for single GPU, I tried to mimic the single-GPU logic as follows. The following updated implementation of `NT_Xent` runs and produces good representations for both single- and multi- GPU:
> 
> ```
> import torch
> import torch.distributed as dist
> from .gather import GatherLayer
> 
> 
> class NT_Xent(torch.nn.Module):
>     
>     def __init__(self, batch_size, temperature, world_size):
>         super(NT_Xent, self).__init__()
>         
>         self.batch_size = batch_size
>         self.temperature = temperature
>         self.world_size = world_size
>         self.full_batch_size = world_size * batch_size
>         
>         self.mask = self.get_self_excluding_mask()
>         self.criterion = torch.nn.CrossEntropyLoss(reduction="sum")
>         self.similarity_f = torch.nn.CosineSimilarity(dim=2)
> 
>     def get_self_excluding_mask(self) -> torch.Tensor:
>         mask = torch.ones((2 * self.full_batch_size, 2 * self.full_batch_size), dtype=bool)
>         mask = mask.fill_diagonal_(0)
>         for i in range(self.full_batch_size):
>             mask[i, i + self.full_batch_size] = 0
>             mask[i + self.full_batch_size, i] = 0
>         return mask
> 
>     def forward(self, z_i: torch.Tensor, z_j: torch.Tensor) -> torch.Tensor:
> 
>         if self.world_size > 1:
>             z_i = torch.cat(GatherLayer.apply(z_i), dim=0)
>             z_j = torch.cat(GatherLayer.apply(z_j), dim=0)
>         
>         # first M entries of z will be the first augmentation; last M entries will
>         # be the second augmentation
>         z = torch.cat((z_i, z_j), dim=0)
>         
>         # the dimension of sim is (2 * self.full_batch_size) x (2 * self.full_batch_size)
>         sim = self.similarity_f(z.unsqueeze(1), z.unsqueeze(0)) / self.temperature
> 
>         sim_i_j = torch.diag(sim, self.full_batch_size)
>         sim_j_i = torch.diag(sim, -self.full_batch_size)
> 
>         positive_samples = torch.cat((sim_i_j, sim_j_i), dim=0).reshape(2 * self.full_batch_size, 1)
>         negative_samples = sim[self.mask].reshape(2 * self.full_batch_size, -1)
> 
>         labels = torch.zeros(2 * self.full_batch_size).to(positive_samples.device).long()
>         logits = torch.cat((positive_samples, negative_samples), dim=1)
>         loss = self.criterion(logits, labels) / (2 * self.full_batch_size)
>         
>         return loss
> ```
> 
> I have changed the order of `GatherLayer` and `torch.cat` operations on `z_i` and `z_j`; I have also changed the positive-pair/self-pair mask.

@mbabadi Hey, have you tested this code in your experiments? I also encountered the same problem, and I think your modification is correct. However, I am wondering about the experimental results in ddp by using this code. Thank you for your attention! Can you please share your insights and experience with it?

**Comment by mbabadi** [comment]
@JUNJIE99 Hey there! So, I didn't run the SimCLR benchmarks on my fix as I was in a rush-- I just compared a short single-GPU run vs. multi-GPU at a comparable number of iterations on my specific data, and they were remarkably close (whereas w/ the original implementation, the loss function wouldn't budge with a multi-GPU run).  For reference, here's what my forked code ended up becoming:
https://github.com/cellarium-ai/SynapseCLR

**Comment by JUNJIE99** [comment]
> @JUNJIE99 Hey there! So, I didn't run the SimCLR benchmarks on my fix as I was in a rush-- I just compared a short single-GPU run vs. multi-GPU at a comparable number of iterations on my specific data, and they were remarkably close (whereas w/ the original implementation, the loss function wouldn't budge with a multi-GPU run). For reference, here's what my forked code ended up becoming:
> https://github.com/cellarium-ai/SynapseCLR

I see! Thanks for taking the time to explain that to me. Thank you!
